lib/aasm/instance_base.rb

```diff
@@ -115,19 +115,25 @@ def may_fire_event?(name, *args)
     end
 
     def fire(event_name, *args, &block)
+      event = @instance.class.aasm(@name).state_machine.events[event_name]
+      raise AASM::UndefinedState, "State :#{event_name} doesn't exist" if event.nil?
+
       @instance.send(event_name, *args, &block)
     end
 
     def fire!(event_name, *args, &block)
-      event_name = event_name.to_s.+("!").to_sym
-      @instance.send(event_name, *args, &block)
+      bang_event_name = "#{event_name}!".to_sym
+      event = @instance.class.aasm(@name).state_machine.events[event_name]
+
+      raise AASM::UndefinedState, "State :#{bang_event_name} doesn't exist" if event.nil?
```

Comment:
I think we can avoid duplicate code from `fire` method

Reply:
@anilmaurya sure. Are you suggesting I extract this into another method then?